core/commands/pin/remotepin.go

```diff
@@ -0,0 +1,694 @@
+		}
+
+		// Execute remote pin request
+		// TODO: fix panic when pinning service is down
+		ps, err := c.Add(ctx, rp.Cid(), opts...)
+		if err != nil {
+			return err
+		}
+
+		// Act on PinStatus.delegates
+		// If Pinning Service returned any delegates, proactively try to
+		// connect to them to facilitate data exchange without waiting for DHT
+		// lookup
+		for _, d := range ps.GetDelegates() {
+			// TODO: confirm this works as expected
+			p, err := peer.AddrInfoFromP2pAddr(d)
+			if err != nil {
+				return err
+			}
+			if err := api.Swarm().Connect(ctx, *p); err != nil {
+				log.Infof("error connecting to remote pin delegate %v : %w", d, err)
+			}
+		}
+
+		// Block unless --background=true is passed
+		if !req.Options[pinBackgroundOptionName].(bool) {
+			requestId := ps.GetRequestId()
+			for {
+				ps, err = c.GetStatusByID(ctx, requestId)
+				if err != nil {
+					return fmt.Errorf("failed to check pin status for requestid=%q due to error: %v", requestId, err)
+				}
+				if ps.GetRequestId() != requestId {
+					return fmt.Errorf("failed to check pin status for requestid=%q, remote service sent unexpected requestid=%q", requestId, ps.GetRequestId())
+				}
+				s := ps.GetStatus()
+				if s == pinclient.StatusPinned {
+					break
+				}
+				if s == pinclient.StatusFailed {
+					return fmt.Errorf("remote service failed to pin requestid=%q", requestId)
+				}
+				tmr := time.NewTimer(time.Second / 2)
+				select {
+				case <-tmr.C:
+				case <-ctx.Done():
+					return fmt.Errorf("waiting for pin interrupted, requestid=%q remains on remote service: check its status with 'ls' or abort with 'rm'", requestId)
+				}
+			}
+		}
+
+		return res.Emit(toRemotePinOutput(ps))
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *RemotePinOutput) error {
+			printRemotePinDetails(w, out)
+			return nil
+		}),
+	},
+}
+
+func multiaddrsToStrings(m []multiaddr.Multiaddr) []string {
+	r := make([]string, len(m))
+	for i := range m {
+		r[i] = m[i].String()
+	}
+	return r
+}
+
+var listRemotePinCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List objects pinned to remote pinning service.",
+		ShortDescription: `
+Returns a list of objects that are pinned to a remote pinning service.
+`,
+		LongDescription: `
+Returns a list of objects that are pinned to a remote pinning service.
+`,
+	},
+
+	Arguments: []cmds.Argument{
+		cmds.StringArg("request-id", false, true, "Request ID of the pin to be listed."),
+	},
+	Options: []cmds.Option{
+		cmds.StringOption(pinNameOptionName, "Return pins objects with names that contain provided value (case-sensitive, exact match)."),
+		cmds.StringsOption(pinCIDsOptionName, "Return only pin objects for the specified CID(s); optional, comma separated."),
+		cmds.StringsOption(pinStatusOptionName, "Return only pin objects with the specified statuses (queued,pinning,pinned,failed)").WithDefault([]string{"pinned"}),
```

Comment:
I'm pretty sure the commands lib doesn't fail in an obvious way if you pass the wrong type of default in here, it's something we probably need to fix.